drivers/i2c/i2c_esp32.c

```diff
@@ -0,0 +1,781 @@
+
+	*config->regs.sda_hold_time = (period / 4) << I2C_SDA_HOLD_TIME_S;
+	*config->regs.sda_sample_time = (period / 4) << I2C_SDA_SAMPLE_TIME_S;
+}
+
+static int i2c_esp32_configure(struct device *dev, u32_t dev_config_raw)
+{
+	const struct i2c_esp32_config *config = dev->config->config_info;
+	struct i2c_esp32_data *data = dev->driver_data;
+	union dev_config dev_config = { .raw = dev_config_raw };
+	int key = irq_lock();
+	u32_t v;
+	int r;
+
+	r = i2c_esp32_configure_pins(config->pins.scl,
+				     config->pins.scl_pull_up,
+				     config->pins.default_level,
+				     config->sig.scl_out,
+				     config->sig.scl_in);
+	if (r < 0) {
+		return r;
+	}
+
+	r = i2c_esp32_configure_pins(config->pins.sda,
+				     config->pins.sda_pull_up,
+				     config->pins.default_level,
+				     config->sig.sda_out,
+				     config->sig.sda_in);
+	if (r < 0) {
+		return r;
+	}
+
+	i2c_esp32_set_device_enabled(config, true);
+
+	v = *config->regs.ctr;
+
+	/* MSB or LSB first is configurable for both TX and RX */
+	if (config->mode.tx_lsb_first) {
+		v |= I2C_TX_LSB_FIRST;
+	} else {
+		v &= ~I2C_TX_LSB_FIRST;
+	}
+
+	if (config->mode.rx_lsb_first) {
+		v |= I2C_RX_LSB_FIRST;
+	} else {
+		v &= ~I2C_RX_LSB_FIRST;
+	}
+
+	if (dev_config.bits.is_master_device) {
+		/* Act as master. */
+		v |= I2C_MS_MODE;
+
+		*config->regs.slave_addr = 0;
+	} else {
+		u32_t addr;
+
+		/* Act as slave. */
+		v &= ~I2C_MS_MODE;
+
+		addr = data->address & I2C_SLAVE_ADDR_V;
+		*config->regs.slave_addr |= addr << I2C_SLAVE_ADDR_S;
+		if (dev_config.bits.use_10_bit_addr) {
+			*config->regs.slave_addr |= I2C_ADDR_10BIT_EN;
+		}
+
+		/* Before setting up FIFO and interrupts, stop transmission */
+		*config->regs.ctr &= ~I2C_TRANS_START;
+
+		/* Byte after address isn't the offset address in slave RAM */
+		*config->regs.fifo_cfg &= ~I2C_FIFO_ADDR_CFG_EN;
+	}
+
+	/* Use open-drain for clock and data pins */
+	v |= (I2C_SCL_FORCE_OUT | I2C_SDA_FORCE_OUT);
+
+	v |= I2C_CLK_EN;
+
+	/* Sample data when SCL is high */
+	v &= ~I2C_SAMPLE_SCL_LEVEL;
+
+	*config->regs.ctr = v;
+
+	switch (dev_config.bits.speed) {
+	case I2C_SPEED_STANDARD:
+		i2c_esp32_configure_frequency(config, 100000);
+		break;
+	case I2C_SPEED_FAST:
+		i2c_esp32_configure_frequency(config, 400000);
+		break;
+	default:
+		r = -EINVAL;
+		goto out;
+	}
+
+	/* Use FIFO to transmit data */
+	*config->regs.fifo_cfg &= ~I2C_NONFIFO_EN;
+
+	v = CONFIG_I2C_ESP32_TIMEOUT & I2C_TIME_OUT_REG;
+	*config->regs.timeout = v << I2C_TIME_OUT_REG_S;
+
+	/* Enable interrupt types handled by the ISR */
+	*config->regs.intr |= I2C_ACK_ERR_INT_ENA_M |
+			      I2C_TIME_OUT_INT_ENA_M |
+			      I2C_TRANS_COMPLETE_INT_ENA_M |
+			      I2C_ARBITRATION_LOST_INT_ENA_M;
+
+	/* Activate the device IRQ */
+	irq_enable(config->irq.line);
+
+out:
+	irq_unlock(key);
+
+	return r;
+}
+
+static inline void i2c_esp32_ack_intrs(const struct i2c_esp32_config *config)
+{
+	/* Acknowledge all interrupts */
```

Comment:
Is this known to be safe?  Often documentation puts stuff like "reserved" in the unused bit fields of an interrupt register and expects you not to ack them spuriously.  Maybe safer to write back the mask of known valid (and handled) interrupts.

Reply:
These are all I2C interrupts, so it's safe.